diff --git a/tools/depends/target/dav1d/DAV1D-VERSION b/tools/depends/target/dav1d/DAV1D-VERSION
index 7897d53556..99b2dba250 100644
--- a/tools/depends/target/dav1d/DAV1D-VERSION
+++ b/tools/depends/target/dav1d/DAV1D-VERSION
@@ -1,5 +1,5 @@
 LIBNAME=dav1d
-VERSION=1.1.0
+VERSION=1.5.0
 ARCHIVE=$(LIBNAME)-$(VERSION).tar.bz2
-SHA512=2eb177b415d1ac7bd08d54b16375afedec61f0771a5181fa93e5e6215856fc26504a7cb47de08abbea5bf18fac5ff91ca31044e7054681c380e0ab152e53aabc
+SHA512=cb6ab1c66acc9f7c792980cbc270e7f3000ecbd4a238b9cf41cc9f557691731f0146bea2483c14a9f70e72611fd1f8c2946faab792c60608ceaf7c0cbfe70b26
 BYPRODUCT=libdav1d.a
diff --git a/tools/depends/target/ffmpeg/FFMPEG-VERSION b/tools/depends/target/ffmpeg/FFMPEG-VERSION
index 6ea8cfb9ad..cf51d8566b 100644
--- a/tools/depends/target/ffmpeg/FFMPEG-VERSION
+++ b/tools/depends/target/ffmpeg/FFMPEG-VERSION
@@ -1,5 +1,5 @@
 LIBNAME=ffmpeg
 VERSION=7.1
 ARCHIVE=$(LIBNAME)-$(VERSION).tar.gz
-SHA512=b0a82ca1a34fb9fa16ee4b7fa682d7c3fdcc68cd703c72487a2de434c714f2dede68d390e61dbb3669e435e271e4580d6bae00875d71a17ad39f43644c5fdd07
+SHA512=3b273769ef1a1b63aed0691eef317a760f8c83b1d0e1c232b67bbee26db60b4864aafbc88df0e86d6bebf07185bbd057f33e2d5258fde6d97763b9994cd48b6f
