diff --git a/tools/depends/target/dav1d/DAV1D-VERSION b/tools/depends/target/dav1d/DAV1D-VERSION
index 7897d53556..99b2dba250 100644
--- a/tools/depends/target/dav1d/DAV1D-VERSION
+++ b/tools/depends/target/dav1d/DAV1D-VERSION
@@ -1,5 +1,5 @@
 LIBNAME=dav1d
-VERSION=1.1.0
+VERSION=1.4.3
 ARCHIVE=$(LIBNAME)-$(VERSION).tar.bz2
-SHA512=2eb177b415d1ac7bd08d54b16375afedec61f0771a5181fa93e5e6215856fc26504a7cb47de08abbea5bf18fac5ff91ca31044e7054681c380e0ab152e53aabc
+SHA512=f624de26a7b7361c29f981a894988b3d70219dc4c0fb3f4e70cd99c5ebee8b64326a4ee88bf7d6138aa66969bb6ab71e3fa37b1f0a13575ff58c42f2ad0befde
 BYPRODUCT=libdav1d.a
diff --git a/tools/depends/target/ffmpeg/FFMPEG-VERSION b/tools/depends/target/ffmpeg/FFMPEG-VERSION
index 6ea8cfb9ad..cf51d8566b 100644
--- a/tools/depends/target/ffmpeg/FFMPEG-VERSION
+++ b/tools/depends/target/ffmpeg/FFMPEG-VERSION
@@ -1,5 +1,5 @@
 LIBNAME=ffmpeg
 VERSION=7.0.2
 ARCHIVE=$(LIBNAME)-$(VERSION).tar.gz
-SHA512=294c411d55d0ef1322d7e882ff0c97d335c9200787ab94ab27ea7a9e9eba272495dedf3a23687271a327fb14e59abc662ef909dd6955bf41a26f2b795196726d
+SHA512=3ba02e8b979c80bf61d55f414bdac2c756578bb36498ed7486151755c6ccf8bd8ff2b8c7afa3c5d1acd862ce48314886a86a105613c05e36601984c334f8f6bf
