diff --git a/tools/depends/target/ffmpeg/FFMPEG-VERSION b/tools/depends/target/ffmpeg/FFMPEG-VERSION
index 6ea8cfb9ad..cf51d8566b 100644
--- a/tools/depends/target/ffmpeg/FFMPEG-VERSION
+++ b/tools/depends/target/ffmpeg/FFMPEG-VERSION
@@ -1,5 +1,5 @@
 LIBNAME=ffmpeg
 VERSION=6.0
 ARCHIVE=$(LIBNAME)-$(VERSION).tar.gz
-SHA512=b3214328f2792364353f38c6b46a71d4970c071d8656b20780abf0e02950167d0933eae825c09102cf8fb19fc679ac444bf1f41a448b624eaa5ea6b0f0bdf4f5
+SHA512=6c19c8aa214194430cd7a3028e86261c1d1aac4cff062e5abd919382729e2a21d26d680e8acaede35e32367c32bcf8872d2e4fb42cbfa9253b2528c197e0b62c
 
